### PR TITLE
Availability after local scheduler failure

### PR DESCRIPTION
**Pull request by stephanie-wang**
Detect and recover from local scheduler failures. The global scheduler updates the db_clients table after a set number of missed heartbeats from a local scheduler. A monitoring script subscribes to these updates and is responsible for marking all tasks that that local scheduler was responsible for as `TASK_STATUS_LOST`.

This does not provide availability if the local scheduler that failed was the one that the driver was connected to.

### CODE REVIEW

File: python/ray/services.py [hunk 1]
```diff
@@ -34,13 +35,14 @@
 # important because it determines the order in which these processes will be
 # terminated when Ray exits, and certain orders will cause errors to be logged
 # to the screen.
-all_processes = OrderedDict([(PROCESS_TYPE_WORKER, []),
+all_processes = OrderedDict([(PROCESS_TYPE_MONITOR, []),
```
Comment (robertnishihara) [inline python/ray/services.py]: When I run `./scripts/start_ray.sh --head` on this branch, the script raises an exception when it exits and prints

```
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/opt/conda/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/home/ubuntu/ray/python/ray/worker.py", line 963, in cleanup
    assert(len(processes) == 0)
AssertionError
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/opt/conda/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/home/ubuntu/ray/python/ray/worker.py", line 963, in cleanup
    assert(len(processes) == 0)
AssertionError
```

Comment (stephanie-wang) [inline python/ray/services.py]: Oh yeah, this is fixed in the next commit.

File: python/ray/worker.py [hunk 2]
```diff
@@ -958,9 +958,13 @@ def cleanup(worker=global_worker):
                               {"end_time": time.time()})
     services.cleanup()
   else:
-    # If this is not a driver, make sure there are no orphan processes.
+    # If this is not a driver, make sure there are no orphan processes, besides
+    # possibly the worker itself.
```
Comment (robertnishihara) [inline python/ray/worker.py]: Why would the worker case be different?

File: src/global_scheduler/global_scheduler.cc [hunk 3]
```diff
@@ -335,6 +388,27 @@ int task_cleanup_handler(event_loop *loop, timer_id id, void *context) {
       free(*pending_task);
     }
   }
+
+  /* Check for local schedulers that have missed a number of heartbeats. If any
+   * local schedulers have died, notify others so that the state can be cleaned
+   * up. */
+  /* TODO(swang): If the local scheduler hasn't actually died, then it should
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.cc]: Why put this in the global scheduler and not the monitor?

File: src/global_scheduler/global_scheduler.h [hunk 4]
```diff
@@ -11,11 +11,15 @@
 /* The frequency with which the global scheduler checks if there are any tasks
  * that haven't been scheduled yet. */
 #define GLOBAL_SCHEDULER_TASK_CLEANUP_MILLISECONDS 100
+#define GLOBAL_SCHEDULER_HEARTBEAT_TIMEOUT 5
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.h]: Can we add a comment saying that `5` refers to the number of missed heartbeats in a row that we're willing to tolerate?

Btw, I suspect this is much too low. This corresponds to a half second delay, and I wouldn't be surprised if that happens frequently.

File: src/global_scheduler/global_scheduler.cc [hunk 5]
```diff
@@ -335,6 +388,27 @@ int task_cleanup_handler(event_loop *loop, timer_id id, void *context) {
...
+  for (int i = utarray_len(state->local_schedulers) - 1; i >= 0; --i) {
+    local_scheduler_ptr =
+        (LocalScheduler *) utarray_eltptr(state->local_schedulers, i);
+    if (local_scheduler_ptr->num_heartbeats_missed >=
+        GLOBAL_SCHEDULER_HEARTBEAT_TIMEOUT) {
+      /* Notify others by updating the global state. */
+      db_client_table_remove(state->db, local_scheduler_ptr->id, NULL, NULL,
+                             NULL);
+      /* Remove the scheduler from the local state. */
+      remove_local_scheduler(state, i);
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.cc]: I wonder if we really want to always remove the local scheduler. What if we only have one local scheduler (e.g., in the single-node case). Then we should probably not remove it in case the missing heartbeats were a false positive.

File: src/global_scheduler/global_scheduler.cc [hunk 6]
```diff
@@ -335,6 +388,27 @@ int task_cleanup_handler(event_loop *loop, timer_id id, void *context) {
...
+   * local schedulers have died, notify others so that the state can be cleaned
+   * up. */
+  /* TODO(swang): If the local scheduler hasn't actually died, then it should
+   * clean up its state and exit upon receiving this notification. */
+  LocalScheduler *local_scheduler_ptr;
+  for (int i = utarray_len(state->local_schedulers) - 1; i >= 0; --i) {
+    local_scheduler_ptr =
+        (LocalScheduler *) utarray_eltptr(state->local_schedulers, i);
+    if (local_scheduler_ptr->num_heartbeats_missed >=
+        GLOBAL_SCHEDULER_HEARTBEAT_TIMEOUT) {
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.cc]: This probably deserves a `LOG_WARN` or something, right?

File: src/global_scheduler/global_scheduler.cc (hunk 5 above)
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.cc]: More generally, if this is the local scheduler on the head node, maybe we don't want to remove it.

File: src/local_scheduler/local_scheduler_client.cc [hunk 7]
```diff
@@ -75,6 +75,7 @@ void local_scheduler_reconstruct_object(LocalSchedulerConnection *conn,
                                         ObjectID object_id) {
   write_message(conn->conn, RECONSTRUCT_OBJECT, sizeof(object_id),
                 (uint8_t *) &object_id);
+  /* TODO(swang): Propagate the error. */
```
Comment (robertnishihara) [inline src/local_scheduler/local_scheduler_client.cc]: We should probably put a fatal check around this (and all other `local_scheduler_client` methods). That can be a different PR though.

File: src/plasma/plasma_store.cc [hunk 8]
```diff
@@ -643,11 +644,23 @@ void send_notifications(event_loop *loop,
...
   }
   /* Remove the sent notifications from the array. */
   utarray_erase(queue->object_notifications, 0, num_processed);
+
+  /* Stop sending notifications if the pipe was broken. */
+  if (closed) {
+    utarray_free(queue->object_notifications);
+    HASH_DEL(plasma_state->pending_notifications, queue);
+    free(queue);
+  }
```
Comment (robertnishihara) [inline src/plasma/plasma_store.cc]: Does it also make sense to close `client_sock`?

File: python/ray/worker.py (hunk 2 above)
Comment (stephanie-wang) [inline python/ray/worker.py]: This is mostly a sanity check that a worker process didn't start any other processes.

File: src/global_scheduler/global_scheduler.cc (hunk 5 above)
Comment (stephanie-wang) [inline src/global_scheduler/global_scheduler.cc]: Yes, we're not handling failure on the head node at all right now. In the future, we should be able to discern between head node and worker node components. It would be a lot to do this as part of this PR though, since handling head node failure means at minimum restarting certain components.

File: python/ray/monitor.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    self.redis = redis.StrictRedis(host=redis_address, port=redis_port, db=0)
+    self.subscribe_client = self.redis.pubsub()
+
+    # Initialize data structures to keep track of the active database clients.
+    self.local_schedulers = set()
+    # Add the NIL_ID so that we don't accidentally mark tasks that aren't
+    # associated with a node as LOST during cleanup.
+    self.local_schedulers.add(NIL_ID)
+
+  def subscribe(self):
```
Comment (pcmoritz) [inline python/ray/monitor.py]: Methods like this should eventually be collected into a python "state" module (I started doing this in experimental/state); I hope once we have full flatbuffer coverage we can use the python flatbuffer module to generate methods that read this state automatically (and get rid of code duplication like DBClientID_SIZE = 20)

File: src/global_scheduler/global_scheduler.h (hunk 4 above)
Comment (stephanie-wang) [inline src/global_scheduler/global_scheduler.h]: Sure. Yeah, I can increase this, say to 10 seconds.

File: python/ray/monitor.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+import logging
+import redis
+import time
+
+from ray.services import get_ip_address
+from ray.services import get_port
+
+# These variables must be kept in sync with the C codebase.
+# common/common.h
+DBClientID_SIZE = 20
```
Comment (pcmoritz) [inline python/ray/monitor.py]: Can we change this to DB_CLIENT_ID_SIZE?

File: src/local_scheduler/local_scheduler_client.cc (hunk 7 above)
Comment (stephanie-wang) [inline src/local_scheduler/local_scheduler_client.cc]: Agreed. It just came up during testing, so I wanted to add the TODO so I didn't forget.

File: src/plasma/plasma_store.cc (hunk 8 above)
Comment (stephanie-wang) [inline src/plasma/plasma_store.cc]: Yes.

File: src/global_scheduler/global_scheduler.cc [hunk 11]
```diff
@@ -167,6 +167,88 @@ void process_task_waiting(Task *waiting_task, void *user_context) {
...
+
+  AuxAddressEntry *entry, *tmp;
+  HASH_ITER(plasma_local_scheduler_hh, state->plasma_local_scheduler_map, entry,
+            tmp) {
+    if (DBClientID_equal(entry->local_scheduler_db_client_id, db_client_id)) {
+      HASH_DELETE(plasma_local_scheduler_hh, state->plasma_local_scheduler_map,
+                  entry);
+      /* The hash entry is shared with the local_scheduler_plasma hashmap and
+       * will
+       * be freed there. */
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.cc]: this line can be combined with the previous

File: python/ray/worker.py (hunk 2 above)
Comment (robertnishihara) [inline python/ray/worker.py]: But why the `<= 1`, why not `== 0`?

File: python/ray/monitor.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    # messages from the subscription channel. This ensures that we start in a
+    # consistent state, since we may have missed notifications that were sent
+    # before we connected to the subscription channel.
+    self.scan_db_client_table()
+    self.cleanup_task_table()
+    log.debug("Scanned schedulers: {}".format(self.local_schedulers))
+
+    # Read messages from the subscription channel.
+    while True:
+      time.sleep(LOCAL_SCHEDULER_HEARTBEAT_TIMEOUT_MILLISECONDS / 1000.)
```
Comment (robertnishihara) [inline python/ray/monitor.py]: we can replace `1000.` with `1000`, since we are doing `from __future__ import division` at the top

File: src/common/redis_module/ray_redis_module.c [hunk 13]
```diff
@@ -972,6 +1048,11 @@ int RedisModule_OnLoad(RedisModuleCtx *ctx,
     return REDISMODULE_ERR;
   }
 
+  if (RedisModule_CreateCommand(ctx, "ray.disconnect", Disconnect_RedisCommand,
+                                "write", 0, 0, 0) == REDISMODULE_ERR) {
```
Comment (robertnishihara) [inline src/common/redis_module/ray_redis_module.c]: hmm, should both this (and `ray.connect`) be `"write pubsub"` instead of just `"write"`?

File: src/common/redis_module/ray_redis_module.c [hunk 14]
```diff
@@ -110,25 +151,60 @@ int Connect_RedisCommand(RedisModuleCtx *ctx,
...
+      OpenPrefixedKey(ctx, DB_CLIENT_PREFIX, ray_client_id, REDISMODULE_WRITE);
+  if (RedisModule_KeyType(db_client_table_key) == REDISMODULE_KEYTYPE_EMPTY) {
+    /* Someone else already deleted this client. */
+    RedisModule_CloseKey(db_client_table_key);
+    RedisModule_ReplyWithSimpleString(ctx, "OK");
+    return REDISMODULE_OK;
+  }
+
+  RedisModuleString *client_type;
+  RedisModuleString *aux_address;
```
Comment (robertnishihara) [inline src/common/redis_module/ray_redis_module.c]: Do we need to call `RedisModule_FreeString` on these two strings?
